### PR TITLE
EPD-983 add ragas evals

### PR DESCRIPTION
**Pull request by adamnolte**

### CODE REVIEW

File: autoblocks/_impl/testing/evaluators/ragas_base.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import abc
+from typing import Generic
+from typing import List
+
+from datasets import Dataset  # type: ignore[import-untyped]
```
Comment (nicolewhite) [inline autoblocks/_impl/testing/evaluators/ragas_base.py:5]: what package is this importing from?

File: autoblocks/_impl/testing/evaluators/ragas_base.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        return Dataset.from_dict(data_dict)
+
+    def get_ragas(self):  # type: ignore[no-untyped-def]
+        # We try to get the openai client first, so that if it doesn't exist we can raise a more informative error
+        # Ragas throws a more generic error if the client or OpenAI api key is not found
+        get_openai_client(evaluator_id=self.id)
+        try:
+            import ragas  # type: ignore[import-untyped]
+        except (ImportError, AssertionError):
+            raise ImportError(f"The {self.id} evaluator requires ragas." "You can install it with `pip install ragas")
```
Comment (nicolewhite) [inline autoblocks/_impl/testing/evaluators/ragas_base.py:72]: nit: make one string

File: autoblocks/_impl/testing/evaluators/ragas_base.py (hunk 1 above)
Comment (adamnolte) [reply autoblocks/_impl/testing/evaluators/ragas_base.py:5]: https://pypi.org/project/datasets/

which comes with ragas, I guess we should import it specifically like Ragas?

File: autoblocks/_impl/testing/evaluators/ragas_faithfulness.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class BaseRagasFaithfulness(BaseRagas[TestCaseType, OutputType], abc.ABC, Generic[TestCaseType, OutputType]):
+    """
+    The RagasFaithfulness evaluator evaluates the factual consistency of the generated answer against the given context.
+
+    See more: https://docs.ragas.io/en/stable/concepts/metrics/faithfulness.html
+    """
+
+    def evaluate_test_case(self, test_case: TestCaseType, output: OutputType) -> Evaluation:
```
Comment (nicolewhite) [inline autoblocks/_impl/testing/evaluators/ragas_faithfulness.py:17]: it looks like `evaluate_test_case` can live on `BaseRagas`, you just need to specify the two things that are changing on each subclass:

```python
class BaseRagasFaithfulness(
  BaseRagas[TestCaseType, OutputType],
  abc.ABC,
  Generic[TestCaseType, OutputType],
):
  ragas_metrics = [ragas.metrics.faithfulness]
  result_name = "faithfulness"

class BaseRagasContextRelevancy(
  BaseRagas[TestCaseType, OutputType],
  abc.ABC,
  Generic[TestCaseType, OutputType],
):
  ragas_metrics = [ragas.metrics.context_relevancy]
  result_name = "context_relevancy"
```

etc

then in `BaseRagas`:

```python
    def evaluate_test_case(self, test_case: TestCaseType, output: OutputType) -> Evaluation:
        dataset = self.make_dataset(test_case=test_case, output=output)
        ragas = self.get_ragas()  # type: ignore[no-untyped-call]
        result = ragas.evaluate(dataset=dataset, metrics=self.ragas_metrics)
        return Evaluation(score=result[self.result_name], threshold=self.threshold)
```

Comment (adamnolte) [reply autoblocks/_impl/testing/evaluators/ragas_faithfulness.py:17]: Is that going to work without importing ragas in the evaluator?

Comment (nicolewhite) [reply autoblocks/_impl/testing/evaluators/ragas_faithfulness.py:17]: oh i see, yeah you could access by name if it's always the same:

```python
class BaseRagasFaithfulness(
  BaseRagas[TestCaseType, OutputType],
  abc.ABC,
  Generic[TestCaseType, OutputType],
):
  metric_name = "faithfulness"
```

```python
        dataset = self.make_dataset(test_case=test_case, output=output)
        ragas = self.get_ragas()  # type: ignore[no-untyped-call]
        result = ragas.evaluate(dataset=dataset, metrics=[getattr(ragas.metrics, self.metric_name)])
        return Evaluation(score=result[self.metric_name], threshold=self.threshold)
```

File: autoblocks/_impl/testing/evaluators/ragas_base.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class BaseRagas(BaseTestEvaluator, abc.ABC, Generic[TestCaseType, OutputType]):
+    """
+    Base class for Ragas Metrics
+    """
+
+    @property
+    @abc.abstractmethod
+    def threshold(self) -> Threshold:
```
Comment (nicolewhite) [inline autoblocks/_impl/testing/evaluators/ragas_base.py:19]: should this be `-> Optional[Threshold]` or is it required

File: tests/util.py [hunk 5]
```diff
@@ -70,3 +49,21 @@ def __repr__(self) -> str:
 
 
 ANY_NUMBER = AnyNumber()
+
+
+class AnyString(str):
+    """
+    Like mock.ANY but checks if the value is any number.
```
Comment (nicolewhite) [inline tests/util.py:56]:
```suggestion
    Like mock.ANY but checks if the value is any string.
```

File: tests/util.py [hunk 6]
```diff
@@ -70,3 +49,21 @@ def __repr__(self) -> str:
...
+    """
+
+    def __eq__(self, other: Any) -> bool:
+        return isinstance(other, str) and not isinstance(other, bool)
```
Comment (nicolewhite) [inline tests/util.py:60]:
```suggestion
        return isinstance(other, str)
```

**Review by nicolewhite — APPROVED** [review]

**Pull request merged**